src/js/actions/layers.js

```diff
@@ -2130,24 +2130,33 @@ define(function (require, exports) {
     selectVectorMask.modal = true;
 
     /**
-     * Dispatches a layer translate for all layers in the document
+     * Dispatches a layer reposition for all layers in the given document model
+     * We may use an outdated document model here, because we want to update the 
+     * positions of layers at the point this event was emitted to us. Please refer to 
+     * https://github.com/adobe-photoshop/spaces-design/pull/2721
+     * and 
+     * https://github.com/adobe-photoshop/spaces-design/issues/2713
      *
+     * @param {Document} document Document model to update
      * @param {object} event Action Descriptor of autoCanvasResizeShift event
      * @return {Promise}
      */
-    var handleCanvasShift = function (event) {
-        var applicationStore = this.flux.store("application"),
-            currentDocument = applicationStore.getCurrentDocument(),
+    var handleCanvasShift = function (document, event) {
+        var layerIDs = collection.pluck(document.layers.all, "id"),
+            newLayers = document.layers.translateLayers(layerIDs, event.to.horizontal, event.to.vertical),
+            positions = newLayers.all.map(function (layer) {
+                return {
+                    layer: layer,
+                    x: layer.bounds ? layer.bounds.left : 0,
+                    y: layer.bounds ? layer.bounds.top : 0
+                };
+            }).toList(),
             payload = {
-                documentID: currentDocument.id,
-                layerIDs: collection.pluck(currentDocument.layers.all, "id"),
-                position: {
-                    x: event.to.horizontal,
-                    y: event.to.vertical
-                }
+                documentID: document.id,
+                positions: positions
             };
-        
-        return this.dispatchAsync(events.document.TRANSLATE_LAYERS, payload);
+    
+        return this.dispatchAsync(events.document.history.optimistic.REPOSITION_LAYERS, payload);
```

Comment:
Using `events.document.history.optimistic.REPOSITION_LAYERS` definitely seems wrong, in that it would create a NEW history state.  It is not clear to me which of the two is correct in your situation:
1) Change to `events.document.history.amendment.REPOSITION_LAYERS` so that this change will be updated on the document model cached as part of the *current* state
2) Change to `events.document.REPOSITION_LAYERS` which would have no history consequences at all.